fix ansible SLES stig remediations in check mode

teacup-on-rockingchair opened the pull request:
#### Description:

- When run in check mode ansible remediation scripts should be more isolated and need extra checks instead of assumptions

#### Rationale:

- Fix ansible remediations run in check mode for rules part of the SLE STIG profile

- Add var_accounts_tmout variable definition for PCI profile used in accounts_tmout rule

marcusburghardt commented:
@teacup-on-rockingchair , could you resolve the conflict in this PR, please?

teacup-on-rockingchair commented:
> @teacup-on-rockingchair , could you resolve the conflict in this PR, please?

Hi @marcusburghardt any issues with the PR that disturbs you?

marcusburghardt commented on linux_os/guide/services/obsolete/r_services/no_host_based_files/ansible/shared.yml:
```diff
@@ -22,6 +22,7 @@
   ansible.builtin.set_fact:
     shosts_equiv_files: '{{ shosts_equiv_files | union(item.stdout_lines) | list }}'
   loop: "{{ result_found_shosts_equiv_files.results }}"
+  when: result_found_shosts_equiv_files is not skipped and item is not skipped
```
There is no condition in the task where the `result_found_shosts_equiv_files` register is created, so it should never be "skipped", making this new criterion not necessary. You could use something like `result_found_shosts_equiv_files.results is defined`. I didn't test this in this case but it should work.

marcusburghardt replied:
@teacup-on-rockingchair , I tried this changes by building the product (e.g. rhel9), copying the generated playbook (build/rhel9/fixes/ansible/no_host_based_files.yml) and executing it in a VM. With or without this new line the result was the same, even in check mode:
`ansible-playbook -i <test vm>, --check no_host_based_files.yml`

```
TASK [Remove Host-Based Authentication Files - Create List of shosts.equiv Files Present in Local File Systems] ******************************************************************************************************************************
ok: [<test vm>] => (item={'changed': False, 'stdout': '', 'stderr': '', 'rc': 0, 'cmd': ['find', '/opt', '-xdev', '-type', 'f', '-name', 'shosts.equiv'], 'start': None, 'end': None, 'delta': None, 'msg': 'Command would have run if not in check mode', 'skipped': True, 'invocation': {'module_args': {'_raw_params': 'find /opt -xdev -type f -name "shosts.equiv"', '_uses_shell': False, 'expand_argument_vars': True, 'stdin_add_newline': True, 'strip_empty_ends': True, 'argv': None, 'chdir': None, 'executable': None, 'creates': None, 'removes': None, 'stdin': None}}, 'stdout_lines': [], 'stderr_lines': [], 'failed': False, 'item': '/opt', 'ansible_loop_var': 'item'})
```
There is no risk of undefined variable in this Playbook as it was designed, so this change is not necessary.

marcusburghardt commented on linux_os/guide/services/obsolete/r_services/no_user_host_based_files/ansible/shared.yml:
```diff
@@ -22,6 +22,7 @@
   ansible.builtin.set_fact:
     user_shosts_files: '{{ user_shosts_files | union(item.stdout_lines) | list }}'
   loop: "{{ result_found_shosts_files.results }}"
+  when: result_found_shosts_files is not skipped and item is not skipped
```
Similar to my previous comment in `no_host_based_files`.

teacup-on-rockingchair replied:
all the conditions about skipped, or not skipper rules is done in order to make sure that when ansible is run in check mode it will not fail. In that case most of the system/cmd steps are skipped, and the patches here are basically to prevent failures to run the ansible playbook.

marcusburghardt replied:
It is not necessary in this rule. It should work fine even in check mode.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/password_expiration/accounts_password_set_max_life_existing/ansible/shared.yml:
```diff
@@ -14,13 +14,13 @@
 {{% endif %}}
   with_items: '{{ user_names.stdout_lines }}'
-  when: user_names.stdout_lines | length > 0
+  when: user_names is not skipped and user_names.stdout_lines | length > 0
```
`user_names is not skipped` should not be necessary since there is no condition in the task where this register is created. Therefore, it will never be "skipped". Maybe would you like to use `user_names.stdout_lines is defined`?

marcusburghardt replied:
@teacup-on-rockingchair unfortunately I couldn't reproduce any error when executing this Playbook in check mode even without the inclusion of `user_names is not skipped and ...`.

I have tested this in RHEL8 and RHEL9, but I believe it should not be different in SLES. Could you give me mode details on how to reproduce the error expected to be fixed by this change, please?

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/password_expiration/accounts_password_set_min_life_existing/ansible/shared.yml:
```diff
@@ -18,6 +18,6 @@
+{{% endif %}}
   with_items: "{{ user_names.stdout_lines }}"
-  when: user_names.stdout_lines | length > 0
+  when: user_names is not skipped and user_names.stdout_lines | length > 0
```
Similar to my previous commend in `accounts_password_set_max_life_existing`.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords_etc_shadow/ansible/shared.yml:
```diff
@@ -13,5 +13,4 @@
   with_items: "{{ users_nopasswd.stdout_lines }}"
-  when: users_nopasswd.stdout_lines | length > 0
-
+  when: users_nopasswd is not skipped and users_nopasswd.stdout_lines | length > 0
```
`users_nopasswd` is not expected to be skipped since there is no condition in the previous task.

marcusburghardt replied:
The command module is skipped in check mode, so it is ok here.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-session/accounts_tmout/ansible/sle12.yml:
```diff
@@ -13,3 +13,4 @@
   file:
     path: /etc/profile.d/autologout.sh
     mode: '0755'
+  when: lookup('fileglob', '/etc/profile.d/autologout.sh', errors='warn')
```
You are not using any wildcard here in the lookup plugin, so it would be better to use `file` instead of `fileglob`, or using another `stat` task.

It is an interesting approach here. Is your intention to explicitly report a "warning" when the file is not present? Otherwise I would recommend to use `stat` module to check the file existence and use its result in the conditional instead of calling the `lookup` directly.
Here is one example: https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/services/ntp/chronyd_or_ntpd_set_maxpoll/ansible/shared.yml#L9

marcusburghardt replied:
Did you have opportunity to check this comment?

marcusburghardt commented on linux_os/guide/system/logging/ensure_rtc_utc_configuration/ansible/shared.yml:
```diff
@@ -5,8 +5,8 @@
+  command: timedatectl set-timezone UTC
   become: yes
-  when: check_tz.rc == 0
+  when: check_tz is not skipped and check_tz.rc == 0
```
I don't think it is relevant to include this new criterion in this Playbook. It won't have any effect. And considering there is already a treatment to fail the playbook, I believe the changes are not necessary.

marcusburghardt commented on linux_os/guide/system/network/network_sniffer_disabled/ansible/shared.yml:
```diff
@@ -13,4 +13,4 @@
     cmd: ip link set dev {{ item.split(':')[1] }} multicast off promisc off
   loop: "{{ network_interfaces.stdout_lines }}"
-  when: "item.split(':') | length == 3"
+  when: network_interfaces is not skipped and "item.split(':') | length == 3"
```
Maybe the intention was to use `network_interfaces.stdout_lines is defined` ?

marcusburghardt commented on linux_os/guide/system/software/integrity/software-integrity/aide/aide_check_audit_tools/ansible/shared.yml:
```diff
@@ -34,9 +34,12 @@
     regexp: ^{{ item }}\s
     line: "{{ item }} {{{ aide_string() }}}"
   with_items: "{{ audit_tools }}"
+  when: "'aide' in ansible_facts.packages"
```
The first task already ensures `aide` is installed so it is not necessary to test it in the next tasks.

marcusburghardt commented on shared/templates/pam_options/ansible.template:
```diff
@@ -33,7 +33,7 @@
     line: '{{{ TYPE }}} {{{ CONTROL_FLAG }}} {{{ MODULE }}}'
     state: present
-  when: '"{{{ MODULE }}}" not in check_pam_module_result.stdout'
+  when: check_pam_module_result is not skipped and '"{{{ MODULE }}}" not in check_pam_module_result.stdout'
```
Would be the intention to use `check_pam_module_result.stdout is defined`?

marcusburghardt reviewed (changes requested):
I have some comments about the new criteria included in the conditionals. Could you take a look, please?

marcusburghardt commented:
In automatus tests for CS8, CS9 and Fedora this error is appearing:
`ERROR - Rule 'network_sniffer_disabled' test setup script 'promisc_interface_exists.fail.sh' failed with exit code 2`
It is likely related to restrictions in containers. Possibly enabling the `CAP_NET_ADMIN` capability should allow the network interfaces inside a container to be set to promiscuous mode but it is not necessary for testing purposes. In VMs, the test scenario works as expected. So the error for this specific test scenario can be safely waived here.

jan-cerny commented:
/packit retest-failed

marcusburghardt commented:
/packit test

teacup-on-rockingchair commented:
> /packit test

Hi @marcusburghardt I cannot seem to understand how the changes made have connection to the failing tests, that are preventing the PR for merging can you give me some hints here, so we can push this one through?

marcusburghardt commented:
> > /packit test
> 
> Hi @marcusburghardt I cannot seem to understand how the changes made have connection to the failing tests, that are preventing the PR for merging can you give me some hints here, so we can push this one through?

Hi @teacup-on-rockingchair , I noticed a weakness in `network_sniffer_disabled` rule that was causing the error while executing the Ansible Playbook in CI tests. It should be fixed by https://github.com/ComplianceAsCode/content/pull/11657 

Once this PR is merged, I would recommend to rebase your PR.

marcusburghardt reviewed (changes requested):
I couldn't reproduce errors in check mode even before the changes. Some changes are interesting in this PR, but I believe most of the changes are not necessary since they have no effect. Maybe you could share more details on how to reproduce the errors in check mode, please?

teacup-on-rockingchair commented:
> I couldn't reproduce errors in check mode even before the changes. Some changes are interesting in this PR, but I believe most of the changes are not necessary since they have no effect. Maybe you could share more details on how to reproduce the errors in check mode, please?

Thanks for the help and feedback @marcusburghardt I rebased at 6abcf65a4740755f4de862e33c04f10a63cc39a5 and removed irrelevant changes, and applied the suggestion to use file instead of fileglob. In my setup the ansible playbooks in check mode now work ok for stig and pci-dss profiles. which was the initial intention.

marcusburghardt commented:
@teacup-on-rockingchair , I am finishing some local tests but so far it is good.
Could you only take a look in the "Automatus SLE15" failure, please? It seems legit:
```
ERROR - Terminating due to error: Couldn't install required packages: passwd.
```

marcusburghardt reviewed (changes requested):
Great. Now it is only necessary to update the references for stability tests and the required tests should pass.

teacup-on-rockingchair commented:
> Great. Now it is only necessary to update the references for stability tests and the required tests should pass.

Thanks @marcusburghardt  :bow: I think the only issue in tests seemed to be network_sniffer_disabled test, which is container related if I understand it correctly

marcusburghardt reviewed (approved):
Thanks

marcusburghardt commented:
Overriding CODEOWNERS as @teacup-on-rockingchair can't approve his own PR.
